src/Microsoft.Azure.WebJobs.Host/Retry/RetryManager.cs

```diff
@@ -0,0 +1,159 @@
+                    Context context = new Context { { "retryCount", 0 }, {"retry", retry} };
+                    await fallbackPolicy
+                        .WrapAsync(GetRetryPolicy(retry))
+                        .ExecuteAsync(async (ctx, tkn) =>
+                        {
+                            int retryCount = (int)ctx["retryCount"];
+                            LogWithId($"Execute with retry, RetryCount: '{retryCount}', Retry: '{retry.Format()}'");
+                            executionResult = await execute(retryCount, executeParams);
+                            RethrowExceptionIfNeeded(executionResult, true);
+                        }, context, token);
+                }
+                else
+                {
+                    LogWithId("Execute without retry");
+                    executionResult = await execute(-1, executeParams);
+                    RethrowExceptionIfNeeded(executionResult, false);
+                }
+            }
+            catch (RetryException retryException)
+            {
+                LogWithId($"Function code returned retry settings: '{retryException.RetryResult.Format()}'");
+                await ExecuteWithRetriesInternalAsync(execute, executeParams, retryException.RetryResult, token);
+            }
+            catch (OperationCanceledException)
+            {
+                LogWithId("Retries were canceled");
+            }
+        }
+
+        private void RethrowExceptionIfNeeded(IEnumerable<FunctionResult> results, bool throwInvocationException)
+        {
+            foreach (var result in results)
+            {
+                if (!result.Succeeded && (result.Exception is RetryException || throwInvocationException))
+                {
+                    throw result.Exception;
+                }
+            }
+        }
+
+        private AsyncRetryPolicy GetRetryPolicy(RetryResult retry)
+        {
+            PolicyBuilder builder = Policy.Handle<FunctionInvocationException>();
+            if (retry.RetryCount == -1)
+            {
+                if (retry.ExponentialBackoff)
+                {
+                    return builder.WaitAndRetryForeverAsync(GetDurationExponential, OnRetryExponentialAsync);
+                }
+                else
+                {
+                    return builder.WaitAndRetryForeverAsync(GetDuration, OnRetryExponentialAsync);
+                }
+            }
+            else
+            {
+                if (retry.ExponentialBackoff)
+                {
+                    return builder.WaitAndRetryAsync(retry.RetryCount, GetDurationExponential, OnRetryAsync);
+                }
+                else
+                {
+                    return builder.WaitAndRetryAsync(retry.RetryCount, GetDuration, OnRetryAsync);
+                }
+            }
+        }
+
+        private TimeSpan GetDuration(int retryCount, Context context)
+        {
+            return ((RetryResult)context["retry"]).SleepDuration;
```

Comment:
are we guaranteed to have this key in the context?

Reply:
yes, all contexts are created with "retry" property